File: src/main/java/com/chattriggers/ctjs/minecraft/libs/renderer/Text.java
package com.chattriggers.ctjs.minecraft.libs.renderer;

import com.chattriggers.ctjs.minecraft.libs.ChatLib;
import lombok.Getter;
import lombok.Setter;
import lombok.experimental.Accessors;
import net.minecraft.client.renderer.GlStateManager;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Used in {@link Renderer#text(String, float, float)}
 */
@Accessors(chain = true)
public class Text {
    /**
     * -- GETTER --
     * Gets the string of the text object
     *
     * @return String value of text object
     *
     * -- SETTER --
     * Sets the string of the text object
     *
     * @param string New value of string
     * @return The Text object for method chaining
     */
    @Getter @Setter
    private String string;

    /**
     * -- GETTER --
     * Gets a list of lines based on the text width
     *
     * @return a list of lines
     */
    @Getter
    private List<String> lines;

    /**
     * -- GETTER --
     * Gets the text color
     *
     * @return A color integer
     */
    @Getter
    private int color;

    /**
     * -- GETTER --
     * Gets if the text is formatted (true by default)
     *
     * @return True if the text is formatted
     */
    @Getter
    private boolean formatted;

    /**
     * -- GETTER --
     * Gets the text x position
     *
     * @return The text x position
     *
     * -- SETTER --
     * Sets the text x position
     *
     * @param x The new text x position
     * @return The Text object for method chaining
     */
    @Getter @Setter
    private float x;

    /**
     * -- GETTER --
     * Gets the text y position
     *
     * @return The text y position
     *
     * -- SETTER --
     * Sets the text y position
     *
     * @param y The new text y position
     * @return The Text object for method chaining
     */
    @Getter @Setter
    private float y;

    /**
     * -- GETTER --
     * Gets the max width for line wrapping
     *
     * @return Max width for line wrapping
     */
    @Getter
    private int width;

    /**
     * -- GETTER --
     * Gets the max number of lines to draw
     *
     * @return Max number of lines to draw
     *
     * -- SETTER --
     * Sets the max number of lines to draw
     *
     * @param maxLines The value of maxLines
     * @return The Text object for method chaining
     */
    @Getter @Setter
    private int maxLines;

    /**
     * -- GETTER --
     * Gets the text scale
     *
     * @return The text scale
     *
     * -- SETTER --
     * Sets the text scale
     *
     * @param scale The new text scale
     * @return The Text object for method chaining
     */
    @Getter @Setter
    private float scale;

    /**
     * -- GETTER --
     * Gets the drop shadow of the text
     *
     * @return Boolean indicating if drop shadow is enabled
     *
     * -- SETTER --
     * Sets the drop shadow
     *
     * @param shadow True to enable drop shadow
     * @return The Text object for method chaining
     */
    @Getter @Setter
    private boolean shadow;

    /**
     * -- GETTER --
     * Gets the text alignment of the text
     *
     * @return String the alignment
     *
     * -- SETTER --
     * Sets the alignment of the text
     *
     * @param align The alignment
     * @return The Text object for method chaining
     */
    @Getter @Setter
    private String align;

    Text(String text, float x, float y) {
        this.string = text;
        this.lines = new ArrayList<>(Collections.singleton(this.string));
        this.formatted = true;
        updateFormatting();
        this.x = x;
        this.y = y;

        this.width = 0;
        this.maxLines = 0;
        this.color = 0xffffffff;
        this.scale = 1;
        this.shadow = false;
        this.align = "left";
    }

    private void updateFormatting() {
        if (this.formatted)
            this.string = ChatLib.addColor(this.string);
        else
            this.string = ChatLib.replaceFormatting(this.string);
    }

    /**
     * Sets if the text is formatted (true by default)
     *
     * @param formatted if the text is formatted
     * @return the Text object for method chaining
     */
    public Text setFormatted(boolean formatted) {
        this.formatted = formatted;
        updateFormatting();
        
        return this;
    }

    /**
     * Sets the text color
     *
     * @see Renderer#color(int, int, int, int)
     * @param color The new color
     * @return The Text object for method chaining
     */
    public Text setColor(int color) {
        this.color = fixAlpha(color);

        return this;
    }

    /**
     * Sets the max width for line wrapping
     *
     * @param width The new max width
     * @return The Text object for method chaining
     */
    public Text setWidth(int width) {
        this.width = width;
        this.lines = Renderer.getFontRenderer().listFormattedStringToWidth(this.string, this.width);

        return this;
    }

    /**
     * Gets whether or not the line wrapped text exceeds the max lines.
     *
     * @return True if wrapped text exceeds max lines
     */
    public boolean exceedsMaxLines() {
        return this.width != 0
                && this.lines.size() > this.maxLines;

    }

    /**
     * Gets the text height based on width.
     *
     * @return The text height
     */
    public int getHeight() {
        return this.width == 0
                ? 9
                :  this.lines.size() * 9;
    }

    public int getMaxWidth() {
        if (this.width == 0) {
            return Renderer.getStringWidth(this.string);
        } else {
            int maxWidth = 0;

            for (String line : this.lines) {
                if (Renderer.getStringWidth(line) > maxWidth)
                    maxWidth = Renderer.getStringWidth(line);
            }

            return maxWidth;
        }
    }

    /**
     * Draws the text onto the client's overlay.
     *
     * @return The text to allow for method chaining
     */
    public Text draw() {
        GlStateManager.enableBlend();
        GlStateManager.scale(this.scale, this.scale, this.scale);
        if (this.width > 0) {
            float maxLinesHolder = this.maxLines;
            float yHolder = this.y;
            for (String line : this.lines) {
                Renderer.getFontRenderer().drawString(line, getXAlign(line), yHolder / this.scale, this.color, this.shadow);
                yHolder += 9;
                maxLinesHolder--;
                if (maxLinesHolder == 0)
                    break;
            }
        } else {
            Renderer.getFontRenderer().drawString(this.string, getXAlign(this.string), this.y / this.scale, this.color, this.shadow);
        }
        GlStateManager.disableBlend();

        Renderer.finishDraw();

        return this;
    }

    // helper method to fix alpha
    private int fixAlpha(int color) {
        int alpha = color >> 24 & 255;
        if (alpha < 10)
            return Renderer.color(color >> 16 & 255, color >> 8 & 255, color & 255, 10);
        return color;
    }

    // helper method to get the x alignment
    private float getXAlign(String string) {
        float x = this.x / this.scale;
        switch (this.align.toLowerCase()) {
            case("center"):
                return x - Renderer.getStringWidth(string) / 2;
            case("right"):
                return x - Renderer.getStringWidth(string);
            case("left"):
            default:
                return x;
        }
    }
}

File: src/main/kotlin/com/chattriggers/ctjs/minecraft/libs/renderer/Text.kt
package com.chattriggers.ctjs.minecraft.libs.renderer

import com.chattriggers.ctjs.minecraft.libs.ChatLib
import com.chattriggers.ctjs.minecraft.objects.display.DisplayHandler
import com.chattriggers.ctjs.utils.kotlin.getOption
import gg.essential.universal.wrappers.message.UTextComponent
import net.minecraft.client.renderer.GlStateManager
import org.mozilla.javascript.NativeObject

class Text {
    private lateinit var string: String
    private var x: Float = 0f
    private var y: Float = 0f

    private val lines = mutableListOf<String>()

    private var color = 0xffffffff
    private var formatted = true
    private var shadow = false
    private var align = DisplayHandler.Align.LEFT

    private var width = 0f
    private var maxWidth = 0
    private var maxLines = Int.MAX_VALUE
    private var scale = 1f

    @JvmOverloads
    constructor(string: String, x: Float = 0f, y: Float = 0f) {
        setString(string)
        setX(x)
        setY(y)
    }

    constructor(string: String, config: NativeObject) {
        setString(string)
        setColor(config.getOption("color", 0xffffffff).toLong())
        setFormatted(config.getOption("formatted", true).toBoolean())
        setShadow(config.getOption("shadow", false).toBoolean())
        setAlign(config.getOption("align", DisplayHandler.Align.LEFT))
        setX(config.getOption("x", 0f).toFloat())
        setY(config.getOption("y", 0f).toFloat())
        setMaxLines((config.getOption("maxLines", Int.MAX_VALUE)).toDouble().toInt())
        setScale(config.getOption("scale", 1f).toFloat())
        setMaxWidth(config.getOption("maxWidth", 0).toInt())
    }

    fun getString(): String = string

    fun setString(string: String) = apply {
        this.string = string
        updateFormatting()
    }

    fun getColor(): Long = color

    fun setColor(color: Long) = apply { this.color = Renderer.fixAlpha(color) }

    fun getFormatted(): Boolean = formatted

    fun setFormatted(formatted: Boolean) = apply {
        this.formatted = formatted
        updateFormatting()
    }

    fun getShadow(): Boolean = shadow

    fun setShadow(shadow: Boolean) = apply { this.shadow = shadow }

    fun getAlign(): DisplayHandler.Align = align

    fun setAlign(align: Any) = apply {
        this.align = when (align) {
            is String -> DisplayHandler.Align.valueOf(align.uppercase())
            is DisplayHandler.Align -> align
            else -> DisplayHandler.Align.LEFT
        }
    }

    fun getX(): Float = x

    fun setX(x: Float) = apply { this.x = x }

    fun getY(): Float = y

    fun setY(y: Float) = apply { this.y = y }

    /**
     * Gets the width of the text
     * This is automatically updated when the text is drawn.
     *
     * @return the width of the text
     */
    fun getWidth(): Float = width

    fun getLines(): List<String> = lines

    fun getMaxLines(): Int = maxLines

    fun setMaxLines(maxLines: Int) = apply { this.maxLines = maxLines }

    fun getScale(): Float = scale

    fun setScale(scale: Float) = apply { this.scale = scale }

    /**
     * Sets the maximum width of the text, splitting it into multiple lines if necessary.
     *
     * @param maxWidth the maximum width of the text
     * @return the Text object for method chaining
     */
    fun setMaxWidth(maxWidth: Int) = apply {
        this.maxWidth = maxWidth
        updateFormatting()
    }

    fun getMaxWidth(): Int = maxWidth

    fun getHeight(): Float {
        return if (lines.size > 1)
            lines.size.coerceAtMost(maxLines) * scale * 10
        else scale * 10
    }

    fun exceedsMaxLines(): Boolean {
        return lines.size > maxLines
    }

    @JvmOverloads
    fun draw(x: Float? = null, y: Float? = null) = apply {
        Renderer.enableBlend()
        Renderer.scale(scale, scale, scale)

        var longestLine = lines.maxOf { Renderer.getStringWidth(it) * scale }
        if (maxWidth != 0)
            longestLine = longestLine.coerceAtMost(maxWidth.toFloat())
        width = longestLine

        var yHolder = y ?: this.y

        for (i in 0 until maxLines) {
            if (i >= lines.size) break
            Renderer.drawString(lines[i], getXAlign(lines[i], x ?: this.x), yHolder / scale, shadow, color)
            yHolder += scale * 10
        }
        Renderer.disableBlend()
        Renderer.finishDraw()
    }

    private fun updateFormatting() {
        string =
            if (formatted) ChatLib.addColor(string)
            else ChatLib.replaceFormatting(string)

        lines.clear()

        string.split("\n").forEach {
            if (maxWidth > 0) {
                //#if MC<=11202
                lines.addAll(Renderer.getFontRenderer().listFormattedStringToWidth(it, maxWidth))
                //#else
                //$$ for (chars in Renderer.getFontRenderer().split(UTextComponent(it).component, maxWidth)) {
                //$$     val builder = ChatLib.TextBuilder(true)
                //$$     chars.accept(builder)
                //$$     lines.add(builder.getString())
                //$$ }
                //#endif
            } else {
                lines.add(it)
            }
        }
    }

    private fun getXAlign(string: String, x: Float): Float {
        val newX = x / scale
        return when (align) {
            DisplayHandler.Align.CENTER -> newX - Renderer.getStringWidth(string) / 2
            DisplayHandler.Align.RIGHT -> newX - Renderer.getStringWidth(string)
            else -> newX
        }
    }

    override fun toString() =
        "Text{" +
                "string=$string, x=$x, y=$y, " +
                "lines=$lines, color=$color, scale=$scale, " +
                "formatted=$formatted, shadow=$shadow, align=$align, " +
                "width=$width, maxWidth=$maxWidth, maxLines=$maxLines" +
                "}"
}
